generator_hdl.py

```diff
@@ -100,63 +95,55 @@ def writeTopModule_interface(topmodule_name, process_list, memories_topin, memor
     # Find names of first & last processing modules in project
     initial_proc = ""
     final_proc = ""
+    notfinal_procs = set()
     for proc in process_list:
-        if proc.is_first: initial_proc = proc.mtype
-        if proc.is_last: final_proc = proc.mtype
+        if proc.is_first: initial_proc = proc.mtype_short()
+        if proc.is_last: final_proc = proc.mtype_short()
+        if extraports and (not proc.is_last):
+            notfinal_procs.add(proc.mtype_short())
 
     string_topmod_interface = writeTopModuleOpener(topmodule_name)
 
     # Write control signals
-    string_ctrl_signals = writeControlSignals_interface(initial_proc, final_proc)
+    string_ctrl_signals = writeControlSignals_interface(initial_proc, final_proc, notfinal_procs)
     
-    # Input arguments
     string_input_mems = ""
-    for memModule in memories_topin:
-        if isinstance(memModule, list): # memories in an array
-            assert(len(memModule)>0)
-            for memModuleInst in memModule:
-                string_input_mems += writeMemoryLHSPorts_interface(memModuleInst)
-        else:
-            string_input_mems += writeMemoryLHSPorts_interface(memModule)
-
-    # Output arguments
     string_output_mems = ""
-    for memModule in memories_topout:
-        if isinstance(memModule, list): # memories in an array
-            assert(len(memModule)>0)
-            for memModuleInst in memModule:
-                string_output_mems += writeMemoryRHSPorts_interface(memModuleInst)
-        else:
-            string_output_mems += writeMemoryRHSPorts_interface(memModule)
-
+    for mtypeB in memDict:
+        memList = memDict[mtypeB]
+        memInfo = memInfoDict[mtypeB]
+        if memList[0].is_initial:
+            # Input arguments
+            string_input_mems += writeMemoryLHSPorts_interface(mtypeB)
+        elif memList[-1].is_final:
```

Comment:
Is there a reason you chose this element in the list? As far as I can tell it shouldn't matter, but maybe I am missing something?

Reply:
Even better, couldn't we use memInfo.is_inital and memInfo.is_final for these lines? For as long as mixedIO isn't supported I think functionally it doesn't matter, but it's maybe a little clearer syntactically what is meant.

Reply:
Good suggestion. Code changed to use memInfo.is_initial & memInfo.is_final. (My original code was indeed a hang-over from me attempting to support chains with mixedIO).